system/acpi.c

```diff
@@ -0,0 +1,345 @@
+            break;
+        }
+        if (memcmp(&config_tables[i].guid, &EFI_ACPI_1_RDSP_GUID, sizeof(efi_guid_t)) == 0) {
+            table_addr = config_tables[i].table;
+        }
+    }
+    return (rsdp_t *)table_addr;
+}
+#endif
+
+static uintptr_t find_rsdp(void)
+{
+    const boot_params_t *boot_params = (boot_params_t *)boot_params_addr;
+
+    const efi_info_t *efi_info = &boot_params->efi_info;
+
+    // Search for the RSDP
+    rsdp_t *rp = NULL;
+#ifdef __x86_64__
+    if (efi_info->loader_signature == EFI64_LOADER_SIGNATURE) {
+        uintptr_t system_table_addr = (uintptr_t)efi_info->sys_tab_hi << 32 | (uintptr_t)efi_info->sys_tab;
+        system_table_addr = map_region(system_table_addr, sizeof(efi64_system_table_t), true);
+        if (system_table_addr != 0) {
+            rp = find_rsdp_in_efi64_system_table((efi64_system_table_t *)system_table_addr);
+            if (rp) rsdp_source = "EFI64 system table";
+        }
+    }
+#else
+    if (efi_info->loader_signature == EFI32_LOADER_SIGNATURE) {
+        uintptr_t system_table_addr = map_region(efi_info->sys_tab, sizeof(efi32_system_table_t), true);
+        if (system_table_addr != 0) {
+            rp = find_rsdp_in_efi32_system_table((efi32_system_table_t *)system_table_addr);
+            if (rp) rsdp_source = "EFI32 system table";
+        }
+    }
+#endif
+    if (rp == NULL) {
+        // Search the BIOS EBDA area.
+        uintptr_t address = *(uint16_t *)0x40E << 4;
+        if (address) {
+            rp = scan_for_rsdp(address, 0x400);
+            if (rp) rsdp_source = "BIOS EBDA";
+        }
+    }
+    if (rp == NULL) {
+        // Search the BIOS reserved area.
+        rp = scan_for_rsdp(0xE0000, 0x20000);
+        if (rp) rsdp_source = "BIOS reserved area";
+    }
+    if (rp == NULL) {
+        // RSDP not found, give up.
+        return 0;
+    }
+    return (uintptr_t)rp;
+}
+
+static uintptr_t find_acpi_table(uint32_t table_signature)
+{
+   rsdp_t *rp = (rsdp_t *)acpi_config.rsdp_addr;
+
+    // Found the RSDP, now get either the RSDT or XSDT
+    // and scan it for a pointer to the table we're looking for
+    rsdt_header_t *rt;
+
+    if (acpi_config.ver_maj < rp->revision) {
+        acpi_config.ver_maj = rp->revision;
+    }
+
+    if (rp->revision >= 2) {
+        rt = (rsdt_header_t *)map_region(rp->xsdt_addr, sizeof(rsdt_header_t), true);
+        if (rt == NULL) {
+            return 0;
+        }
+        // Validate the XSDT.
+        if (*(uint32_t *)rt != XSDTSignature) {
+            return 0;
+        }
+        rt = (rsdt_header_t *)map_region(rp->xsdt_addr, rt->length, true);
+        if (rt == NULL || acpi_checksum(rt, rt->length) != 0) {
+            return 0;
+        }
+        // Scan the XSDT for a pointer to the table we're looking for.
+        uint64_t *tab_ptr = (uint64_t *)((uint8_t *)rt + sizeof(rsdt_header_t));
+        uint64_t *tab_end = (uint64_t *)((uint8_t *)rt + rt->length);
```

Comment:
What is a reasonable upper bound for rt->length ? Certainly much lower than UINT32_MAX or even 1 GB.

Reply:
Probably, but that was set according to ACPI specs.